core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -144,8 +152,16 @@ public void close() throws IOException {
     // org.apache.hadoop.fs.FileSystem.close may check the existence of certain temp files before
     // closing
     super.close();
-    if (mContext != null && mContext != FileSystemContext.get()) {
-      mContext.close();
+    synchronized (CACHED_CONTEXTS_LOCK) {
+      Pair<FileSystemContext, AtomicInteger> contextAndRefCount =
+          CACHED_CONTEXTS.get(mContext.getParentSubject());
+      if (contextAndRefCount == null) {
+        LOG.warn("Failed to find context to close, have you called close multiple times?");
+        return;
+      }
+      if (contextAndRefCount.getSecond().decrementAndGet() == 0) {
```

Comment:
can we push the reference counting and close implementation into `FileSystemContext`? It reads better than `getFirst()/getSecond()`.

Reply:
If we move the caching to FSC I think this makes sense, otherwise I think the ref counting should be done here.